Mathlib/Analysis/SumOverResidueClass.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Data.ZMod.Basic
+import Mathlib.Topology.Instances.ENNReal
+
+/-!
+# Sums over residue classes
+
+We consider infinite sums over functions `f` on `ℕ`, restricted to a residue class mod `m`.
+
+The main result is `summable_indicator_mod_iff`, which states that when `f : ℕ → ℝ` is
+decreasing and takes nonnegative values, then the sum over `f` restricted to any residue class
+mod `m ≠ 0` converges if and only if the sum over all of `ℕ` converges.
+
+(Note that we want to use `Summable.of_nonneg_of_le`, which insists on `ℝ` as the target.)
+-/
+
+
+open BigOperators in
+lemma Finset.sum_indicator_mod {R : Type*} [AddCommMonoid R] (m : ℕ) [NeZero m] (f : ℕ → R) :
+    f = ∑ a : ZMod m, {n : ℕ | (n : ZMod m) = a}.indicator f := by
+  ext n
+  simp only [Finset.sum_apply, Set.indicator_apply, Set.mem_setOf_eq, Finset.sum_ite_eq,
+    Finset.mem_univ, ↓reduceIte]
+
+variable (m : ℕ) [hm: NeZero m]
+
+open Set in
+/-- A sequence `f` with values in an additive topological group `R` is summable on the
+residue class of `k` mod `m` if and only if `f (m*n + k)` is summable. -/
+lemma summable_indicator_mod_iff_summable {R : Type*} [AddCommGroup R] [TopologicalSpace R]
+    [TopologicalAddGroup R] (k : ℕ) (f : ℕ → R) :
+    Summable ({n : ℕ | (n : ZMod m) = k}.indicator f) ↔ Summable fun n ↦ f (m * n + k) := by
+  trans Summable ({n : ℕ | (n : ZMod m) = k ∧ k ≤ n}.indicator f)
+  · rw [← (finite_lt_nat k).summable_compl_iff (f := {n : ℕ | (n : ZMod m) = k}.indicator f)]
+    simp only [summable_subtype_iff_indicator, indicator_indicator, inter_comm, setOf_and,
+      compl_setOf, not_lt]
+  · let g : ℕ → ℕ := fun n ↦ m * n + k
+    have hg : Function.Injective g := fun m n hmn ↦ by simpa [g, hm.ne] using hmn
+    have hg' : ∀ n ∉ range g, {n : ℕ | (n : ZMod m) = k ∧ k ≤ n}.indicator f n = 0 := by
+      intro n hn
+      contrapose! hn
+      exact (Nat.range_mul_add m k).symm ▸ mem_of_indicator_ne_zero hn
+    convert (Function.Injective.summable_iff hg hg').symm using 3
+    simp only [Function.comp_apply, mem_setOf_eq, Nat.cast_add, Nat.cast_mul, CharP.cast_eq_zero,
+      zero_mul, zero_add, le_add_iff_nonneg_left, zero_le, and_self, indicator_of_mem, g]
+
+variable {f : ℕ → ℝ} (hf : Antitone f) (hf₀ : ∀ n, 0 ≤ f n)
```

Comment:
Do you really need `hf0`? If it is not true, then nothing is ever convergent, so all your lemmas should still be true, right? By the way, could you put the Prop assumptions inside the lemmas instead of as global variables? Otherwise, it makes the file hard to read as the assumptions are not obvious.

Reply:
I'll move the Prop assumption inside the lemmas. You are right about `hf₀`; I'll work out how to remove it.